### ISSUE TITLE
How to distinguish between spheres which are created using "addSphere()" and "state.currentGroup"?

### ISSUE THREAD

**Issue by simonryu** [molstar/molstar#579]
I created three groups where each group consists of two spheres using [this code](https://github.com/molstar/molstar/issues/535#issuecomment-1259445564) as follows (refer to the figure below). 
Group 1 (two red spheres), Group 2 (two green spheres), and Group 3 (two blue spheres). 

```
const state = MeshBuilder.createState(data.centers.length * 128, data.centers.length * 128 / 2, mesh);
state.currentGroup = 0;
for(let i=0, j=0; i < data.centers.length; i +=3, ++j){
    state.currentGroup = j;
    addSphere(state, Vec3.create(data.centers[i],data.centers[i+1],data.centers[i+2]), data.radii[j], 3);
}
MeshBuilder.getMesh(state);
```

Because I assigned an unique "state.currentGroup" to each sphere as in the code above, my code worked fine to capture the interaction of each sphere. 

My questions are as follows.

**Q1)** How can I distinguish between **spheres in the same group** using some kind of "Id", e.g. state.currentGroup? I.e. is there a way either to use state.currentGroup of a sphere or to store an "ID" for the sphere so that my program identifies clicked sphere?

**Q2)** How can I determine first two arguments "initialCount" and "chunkSize" of MeshBuilder.createState()? Currently, I just used values of example code in the repository of Molstar.

<img width="1332" alt="스크린샷 2022-10-06 오후 10 06 19" src="https://user-images.githubusercontent.com/9817432/194322727-3b7c45d6-0655-4aa6-9aff-3c258510087a.png">

**Comment by dsehnal** [comment]
> **Q1)** How can I distinguish between **spheres in the same group** using some kind of "Id", e.g. state.currentGroup? I.e. is there a way either to use state.currentGroup of a sphere or to store an "ID" for the sphere so that my program identifies clicked sphere?
> 

You cannot distinguish between things with the same group id in the same visual.

> **Q2)** How can I determine first two arguments "initialCount" and "chunkSize" of MeshBuilder.createState()? Currently, I just used values of example code in the repository of Molstar.

That is the estimate of how many vertices your mesh will have and how much should the buffer grow by when your original estimate is too low.

**Comment by simonryu** [comment]
@dsehnal Thank you for your quick reseponse. I copy some code chunk below from [here](https://github.com/molstar/molstar/issues/535#issuecomment-1259445564) to make sure about your answer to Q1).

My question was how I can distinguish spheres with different "state.currentGroup" values in the same "spheresGroup" which comes from "structure.apply(StateTransforms.Misc.CreateGroup, ...)" below.

In this case, is your answer to Q1) still that I cannot distinguish spheres?

```
const structure = this.plugin.build().toRoot();
const spheresGroup = structure.apply(StateTransforms.Misc.CreateGroup, { label: 'SphereGroup' });
for(let i = 0;i < numberOfSphereSet; ++i)
{
      spheresGroup.apply(CreateSpheres, {    
      index  : i+1,
      centers: spheres.map(a=>a.get_center()).flat(),
      radii  : speheres.map(a=>a.m_radius),
      sphereColor: ColorNames.red
      });
}
await structure.commit();
```

If your answer to Q1) is the same, please provide me with answer to Q1') below.

Q1') Let SG_i, i=1,2,3,...,6 be six different sphereGroups. Consider that I apply a sphere to a SG_i. Then we may distinguish those six spheres (right?). Let SSG_j, j=1,2,3 be another three different sphereGroups. Can I add SG_1 and SG_2 to SSG_1 and manage two spheres in a group? Similarly, can I add SG_3 and SG_4 to SSG_2? SG_5 and SG_6 to SSG_3?

**Comment by dsehnal** [comment]
The event also contains the information about the representation you clicked on, you can distinguish between the spheres like that, by including some more info in the `sourceData` field (or something like this) when you are creating your shape.

**Comment by simonryu** [comment]
@dsehnal Thanks. I will try.

**Comment by simonryu** [comment]
<Figure 1. Click events of two spheres in a sphereGroup>
![스크린샷 2022-10-08 오후 1 03 46](https://user-images.githubusercontent.com/9817432/194687484-9f084fb0-8403-443b-9238-d3efd227ee51.png)

**Comment by simonryu** [comment]
<Figure 1. Click events of two spheres in a sphereGroup>
![image](https://user-images.githubusercontent.com/9817432/194687552-e7248df4-89b1-4ccd-b137-956332e070c9.png)

**Comment by simonryu** [comment]
![스크린샷 2022-10-08 오후 1 03 46](https://user-images.githubusercontent.com/9817432/194687599-e039fd7d-cc23-415c-b921-7d909a95a239.png)

**Comment by simonryu** [comment]
<Figure 1. Click event of two sphres in a sphereGroup>

![ClickEvent1](https://user-images.githubusercontent.com/9817432/194687838-783b9445-e8da-4383-9902-2d86b5b2b234.png)

**Comment by simonryu** [comment]
@dsehnal 
I have checked out "sourceData" via developer tools of Chrome browser. 

Figure 1 shows the result of two clicks of two spheres (Sphere 1 and Sphere 2) of the same sphereGroup. Figure 2 shows the sourceData of Sphere 1. 

Both Sphere 1 and 2 have the same sourceData because a **sourceData** is not available per a sphere but **per a sphereGroup (shape)** 

NOTE: In this case, my sphereGroup has 33 spheres. 

Q) I can define a different id for each sphere when I create it. **Where can I include this information** so that my program identify which sphre is clicked?

**<Figure 1>**
![ClickEvent1](https://user-images.githubusercontent.com/9817432/194691309-d1339aa3-7cc8-4779-9af6-79d9848a36ec.png)

**<Figure 2>**
![ClickEvent2](https://user-images.githubusercontent.com/9817432/194691342-9d227a65-7a67-4197-a895-9e1a8c5a4649.png)

**Comment by dsehnal** [comment]
You cannot distinguish between two spheres with the same group id inside the same visual. You need to give them different group id and keep a mapping outside of the visual itself (can be in source data I guess too). 

These are no longer Mol* questions but you are essentially asking for help with building the base logic of your app. Please try to spend more time next time trying to figure out these problems yourself before you ask.

**Comment by simonryu** [comment]
@dsehnal 
Sorry for bothering you. I have spent much time but still not familiar with using molstar. Please understand my situation. 

It would be more efficient to tell you what I wanted my code to do as follows.

1) A sphere belongs to a group ("Group 1", "Group 2", or "Group 3" in state tree of [this figure](https://github.com/molstar/molstar/issues/579#issue-1399647512)).

2) A group can have more than two spheres.

3) All groups belong to a (topmost) single group ("Groups" in state tree of [this figure](https://github.com/molstar/molstar/issues/579#issue-1399647512)). 

4) When a user hovers a group lable (e.g. "Group 1", "Group 2", or "Group 3") in state tree, program shows the label of group. 

5) When a user click a sphere, program identifies radius of the clicked sphere.

**My current program works fine for 1) through 4). However, it does not work for 5).** 

As shown in [the code snippet](https://github.com/molstar/molstar/issues/579#issuecomment-1270455023), current code creates a topmost group and then repeatedly calls "apply" for a set of spheres using the topmost group.

Are you advising that I should assign an id to each sphere when I create a set of spheres?

Could you give me more hints?

**Comment by simonryu** [comment]
I created a set spheres with different state.currGroup values as in [the code snippet](https://github.com/molstar/molstar/issues/579#issue-1399647512) and called Shape.create().

**Comment by dsehnal** [comment]
The group id is not the same as a state tree node. If you have 10 spheres as part of single state tree node and you need to distinguish between them, you need to allocate 10 unique group ids to these 10 spheres within the single representation.

**Comment by simonryu** [comment]
@dsehnal 
Thank David for your help. Now, my code works fine. 

I created an array "InfoMap" of string in sourceData which stores an information to show when a sphere is clicked. 
Then I added  a callback function as one of arguments for Shape.create(...) as in [this function](https://github.com/molstar/molstar/blob/2ee0f3bf97031a58ed5acae14445b23647696dca/src/tests/browser/render-shape.ts#L97).

```
Shape.create(
...
(groupId: number) => 'Info: ' + InfoMap[groupId],
...
);
```

**Comment by simonryu** [comment]
![스크린샷 2022-10-13 오전 9 20 38](https://user-images.githubusercontent.com/9817432/195471097-8d91447b-b88d-4c22-953c-ed57f295c405.png)

**Comment by simonryu** [comment]
I store the information in a string array (named "InfoMap") when I create each sphere in advance. Then molstar shows the label (in this example radius of the sphere) in InfoMap which corresponds to a clicked sphere as illustrated in the figure below.

![ShowRadius](https://user-images.githubusercontent.com/9817432/195472368-6d86e2c1-98dd-439f-ac04-84898c2fcc2c.png)

@dsehnal 
Is there a way to identify "the id (or label whatever)" of the clicked sphere and then to process something using "the id" rather than using the information stored in advance when I create the sphere?

I tried to identify the label of the clicked sphere using the following code.

```
this.plugin.behaviors.interaction.click.subscribe(e => {
            type Loci = {
                [key: string]: any;
            };
            const lc : Loci = e.current.loci;
            console.log(lc.shape.getLabel());
}
...
Shape.create(
...
(groupId: number) => 'ID: ' + SphereIDMap[groupId],
...
);
``` 
I stored a globally unique id in SphereIDMap rather than an unique in the same representation (shape).
I expected that "shape.getLabel()" shows the globally unique id of the clicked sphere. However, the console shows "ID: undefined" only.
